### DISCUSSION TITLE
Issue with DTD Validation : tei_all.dtd and tei_lite.dtd

### DISCUSSION THREAD

**Discussion by mbodjabdoulaziz** [TEIC/TEI#2585] [Q&A] [answered]
Hello,

A researcher wants to make modification to a TEI document but is unsure about the validation of the file and asked me to check.
Using libxml2, I have multiple error during the validation process but the document is marked as valid in the Oxygen software used by the researcher. Reverting the xml document to its original state (before the researcher's modification) makes it valid using libxml2 and the dtd file tei_all.dtd generated using https://roma.tei-c.org/.

The researcher added hundreds of `<choice>` in the document after consulting these pages : https://www.tei-c.org/release/doc/tei-p5-doc/en/html/examples-choice.html and https://www.tei-c.org/release/doc/tei-p5-doc/en/html/ref-choice.html.

Here is one exemple :  `<choice><expan>pre<ex>n</ex>di</expan><abbr>prēdi</abbr></choice>`

Here is the error message from the validation :

`Element choice content does not follow the DTD, expecting (sic | corr | reg | orig | unclear | abbr | expan | ex | am | supplied | seg | choice), got (expan abbr) `

I used the exemple from this page https://www.tei-c.org/release/doc/tei-p5-doc/en/html/ref-choice.html and the validation still returned some errors but I managed to avoid validation issues making a small change : 

 - FROM
~~~xml
<choice>
  <orig>favour</orig>
  <reg>favor</reg>
 </choice>
~~~

- TO
~~~xml
<choice>
  <orig>favour  <reg>favor</reg></orig>
 </choice>
~~~

I decided to open a discussion in order to have advice for this researcher. How to use this `<choice>` element in the right way ?
Thank you.

**Comment by sydb** [comment]
On quick read, I think y’all are using `<choice>` correctly, which makes me think either you are not using the DTD correctly or there is a bug in the DTD.
I will start taking a look at this now, but may well go to bed before I get to any real answers.

**Comment by sydb** [comment, marked as answer]
Confirmed. This is a bug in the TEI’s tei_all.dtd file, probably caused by an error in the Stylesheets. My suspiscion (without having looked at the code) is that it will be fixable, but not easy. But even if it gets fixed next week, it may not make it into a release until January.

May I ask why you and the researcher are using DTDs at all? In general it is recommended to use RELAX NG plus Schematron for validating TEI, rather than DTDs (or even XSDs) plus Schematron, because quite a few of the constraints expressed in the TEI RELAX NG schemas cannot be expressed in the DTD language. (And although libxml2 can validate against RELAX NG — e.g., `xmllint --relaxng /path/to/tei_all.rng /path/to/file.xml` — I would strongly recommend `jing` instead. My experience is that although `xmllint` gives the right answer as to whether or not a file is valid, the error messages are inscrutable, or outright misleading. If you don’t like `jing` (which is what oXygen uses, I think) you might try `rnv`.)

If for some reason you need to continue using DTDs, it is possible to re-define `<choice>` so that it works. The current content model for `<choice>` is
~~~xml
  <content>
    <alternate minOccurs="2" maxOccurs="unbounded">
      <classRef key="model.choicePart"/>
      <elementRef key="choice"/>
    </alternate>
  </content>
~~~
It is the `minOccurs="2"` that is causing the problem. Thus if you override this declaration with
~~~xml
  <content>
    <sequence>
      <alternate minOccurs="1" maxOccurs="1">
        <classRef key="model.choicePart"/>
        <elementRef key="choice"/>
      </alternate>
      <alternate minOccurs="1" maxOccurs="unbounded">
        <classRef key="model.choicePart"/>
        <elementRef key="choice"/>
      </alternate>
    </sequence>
  </content>
~~~ 
the DTD that is generated will have a working content model, at least for `<choice>`. The same problem is likely to occur for `<domains>`, `<targFunc>`, `<points>`, `<transpose>`, `<target>`, `<weights>`, and `<modelSequence>`. However, I tried this myself and `xmllint` had other complaints about the DTD. Way too tired to chase those now, though.

**Reply by mbodjabdoulaziz** [reply]
Thank you for your answer. 

I wrote a program to help this researcher. The program is loaded the xml file in a database and then created a new xml file. This helped the researcher a lot but at that point the changes were only in the data. 

Later the researcher came back with those modifications that I posted and as I talked about the validity of the xml, the researcher asked me to check that. It's my first time ever working with TEI so I just downloaded the tei_all.dtd because it was available on the website.

I followed your advices and using jing and the tei_all.rng file I can validate the document which is what the researcher wants. I also did look into rnv but could not build the program so I will go ask for information on their github. 

I closed my question as everything is fine for me now.
